Separate scripts and reports

jonasfreimuth opened the pull request:
Additional changes
* remove sigmut db
* format scripts when touched
* update test/settings.yaml with current database locations
* convert file io to use CSV files wherever possible

jonasfreimuth committed:
style reports

* qc_report_per_sample.Rmd & taxonomic_classification.Rmd

* apply styler tidyverse style
* format markdown to fit line widths

jonasfreimuth committed:
simplify chunk opts in reports

* qc_report_per_sample.Rmd & taxonomic_classification.Rmd

jonasfreimuth committed:
remove sigmut_db

* done as it was no longer in use
* formated render_variant_report rule during this

jonasfreimuth committed:
isolate computation code from variantreport

* WIP isolate computation code from variantreport
* simplify function in mutation_parsing_funs.R
* rename deconvolution functions file
* rename deconvution runner script
* add saving and loading of sigmut dfs
* add saving and loading for variant abundance df
* move data dl generation back to variant report
* reenable executeDeconvolution control
* simplify code echo setting in variantreport
* merge function files necessary for deconvolution.R
* remove reference to unmerged deconvolution funs
    also format a bit
* add command line arguments to deconvolution script
* move command line parsing in deconvolution script
* adapt params of variantreport; set default to ""
* simplify parameter handling in variantreport rmd
* update snakefile to split deconvolution & report
* fix filename of variant abundance file in deconvolution
* add dummy outputs for deconvolution
* simplify saving of outputs of deconvolution
* fix render_variant_report rule
* fix file io related issues in deconvolution
* add logging for file io in deconvolution
* that might be an overstatement
* add missing package to variantreport
* add missing parameters for variantreport
* add logging of used parameters for report rendering
* simplify sample_sheet parsing
* format deconvolution, add todos
* rename most vars not following conventions
* separate variant pooling into own file
* simplify sigmut preprocessing
* move bulk_freq_exctraction to appropriate place
* finish renaming variant col
* slightly simplify sigmat creation
* style deconvoltion & add comments
* add package loading to deconvolution_funs.R
* change deconvolution dummy file with 0 rows

jonasfreimuth committed:
convert all file io to csv where possible

jonasfreimuth committed:
style recently modified files

jonasfreimuth committed:
fix empty files being read in mut/var summary

jonasfreimuth committed:
fix typo in error message for empty files

jonasfreimuth committed:
remove leftover function script

jonasfreimuth committed:
add new scripts to Makefile.am

jonasfreimuth committed:
add formatted printing of params to simplify debugging

jonasfreimuth committed:
restore code of orig sep scripts rep branch

* in all affected files

jonasfreimuth committed:
restore format of orig sep scripts rep branch

jonasfreimuth committed:
remove useless variant pooling script

jonasfreimuth commented:
`make distcheck` completes successfully.

alexg9010 commented on scripts/Sample_prep_scripts/sigmuts_on_amplicons.rmd:
```diff
@@ -34,7 +34,7 @@ amplicon_gr <- makeGRangesFromDataFrame ( amplicon, keep.extra.columns = TRUE )
 # TODO: make the mutation sheet with nt position for ALL mutatation and loop through it
 # mut <- read.csv("../tests/mutation_sheet_detailed.csv", header = TRUE ) # TODO provide as input
-mut <- read.table("../tests/sample_data/signature_mutations/B117_outbreakinfo_mutation_report_data_2021-03-06.tsv", header = TRUE)
+mut <- read.csv("../tests/sample_data/signature_mutations/B117_outbreakinfo_mutation_report_data_2021-03-06.tsv", header = TRUE)
```
Why would you use the `read.csv` function to read a tab-separated file?
I would either add the right parameters (separators, headings) to the `read.table` call or change to `read.delim`, which is the function to read tsv files.

alexg9010 commented on scripts/creating_mutation_summary_table.R:
```diff
@@ -28,14 +31,14 @@ create_summary <- function ( files, output_file ){
 }
 
 args <- commandArgs (trailingOnly=TRUE)
+#args <- c("/home/jfreige/proj/pigx_sars-cov-2/tests/output/mutations", "/home/jfreige/proj/pigx_sars-cov-2/tests/output/variants/data_mutation_plot.csv")
```
This should not be part of the PR.

alexg9010 commented on scripts/creating_variants_summary_table.R:
```diff
@@ -13,12 +14,15 @@ create_summary <- function ( files ){
 
+  # remove empty files from list
+  variants_list_has_rows <- sapply(variants_list, nrow)
+  variants_list <- variants_list[variants_list_has_rows > 0]
```
If you don't reuse the `variants_list_has_rows` elsewhere, I suggest you combine the two lines into one.

jonasfreimuth replied:
Done

alexg9010 commented on scripts/fun_index.R:
```diff
@@ -0,0 +1,53 @@
+  capture.output(write.csv(data,
+    file = "",
+    quote = FALSE, row.names = rownames
+  )) %>%
```
To use the pipe you also need to load `library(dplyr)` here, or you rewrite this function without the pipe.

jonasfreimuth replied:
package should be loaded now

jonasfreimuth committed:
isolate code from index.Rmd

* style index.Rmd
* isolate code from index.Rmd
* move index.Rmd computations to separate script
* add function file for index
* remove unneeded pkgs from index.R
* add cond running of mutation regression in index.R
* make index.Rmd chunk names more descriptive
* style index.Rmd
* prune parameters for index.Rmd
* prune parameters for index.R
* add saving of lin reg results
* move threshold formatting back to report
* add data loading scaffold to index.Rmd
* move a plot data generation step back to report
* add pooling weights calculation to report
* move mutation counting to script
* move mutation filtering & pooling to report
* simplify conditional execution switches in script
* add creation of mutations_sig df to report
* add loading of unfiltered mutations_sig to report
* fix conditional execution of mutation counts
* move generation of mut & var plot dfs to report
* move weights calculation to fit with old flow
* move quality filtering back to report
* add loading of mutations count df to report
* cleanup and style index.R
* style and error correct index.Rmd
* add calculation of good_samples_df to script
* remove unnecessary loading of variants file
* rename mutation regression script
* add arguments to mutation_regression.R
* change mut reg script to get filenames from params
* add snakefile rule for mutation_regression.R
* fix interaction of snakefile & mutation_regression
* fix mutation_regression empty file output
* add calculation of cond execution vars to report
* rename index functions file for consistency
* fix conditional execution in index.Rmd
* add missing parameters to index.Rmd & snakefile
* add package loading to fun_index.R

jonasfreimuth committed:
remove unused mutation sheets

jonasfreimuth committed:
simplify deconvolution

  simplify parse_snv_csv function

  remove data frame coercion of parsed snv csv

  remove unused script vep_rule.py

  remove reference to vep_rule.py in Makefile.am

  add comments and TODOs to deconvolution.R

  simplify creation of complete_df

  correct coverage to depth

  remove dep filtering TODO again

  fix leftover cov to dep

  fix mut read depth filter

    the parameter is a string and so needs conversion before
    the comparison

  fix coverage -> depth issue in variantreport

  add specification of read depth threshold settings

  add forgotten mut dep threshold to run_deconv rule

  remove depth filter via settings TODO

  fix variable names & some formatting

  remove outcommented code

  change variable names to snake_case

  also some formatting changes bc I was lazy

  change deconv function names to snake_case

  fix variable name

  change variable names to snake_case

  fix formatting in deconvolution_funs.R

  format deconvolution.R

  simplify create_sig_matrix function

  add rownames to sig_mat

  correct step numbering

  add explicit package to function

  remove unneeded 3rd output of simulate_others

  format simulate_others function

  add additional explanation for simulate_others

  simplify sig creation

  simplify msig_stable_all creation

  WIP simplify simulate_others

  remove muts col from simulate_others output

  create msig_simple_unique without muts col

  massively simplify signature matrix deduplication

  remove now unnecessary dedupe functions

  use var_sep in all appropriate places

  add missing var_sep use

  add sigmut coercion to data frame

  for easier handling

  fix col comparison for dataframe

  use named list for saving of dupe group names

  very slightly simplify col renaming

  fix sigmut_deduped_df variable name

  simplify all_variants creation

  resolve TODO

  comment proportion weights creation

  simplify msig weighting

  fix variable naming

      vektor -> vector

  simplify weighting of inverted matrix

  remove obsolete filtering of muts col

  adapt variable naming in simulate_others

  move code from simulate_others into script

    the code now is not too complicated, this function is only used once
    and this is a rather critical step, so the code should be in the
    script for better compehension

  remove vector coercion in devonv

    this should preserve naming of abundance values, giving their
    variant

  change pound sign to n in comments

    hopefully improves readability

  simplify creation of variant_abundance_df

  complete renaming of var abundance df

  make case checking for group splitting more robust

    now uses the full precision available on a given machine

  simplify group splitting case 1

  fix variables

  massively simplify dupe group splitting

  fix output of dummy variants file

  invert execute_deconvolution condition

  fix log message for dummy variants file

  fix dummy output for variants with meta file

  massively simplify creation of output_variant_plot

  simplify mut frequency extraction

  simplify output_mutation_frame creation

  tentatviely remove coercion to character

    I have no earthly clue why we would need that.
    but if we do i hope i remembere i took it out

  add TODO for filtering muts

  simplify get_protein_mut

    introduce a rather large TODO: resolve deletions business
    and adapt variable naming

  simplify creation of output_mutation_frame a bit

    also takes care of variable naming issues

  add filtering of aa string containing NA

    add TODO for finding out why that happens

  fix variable naming in variant report

    leftovers from deconvolution cleanup

  rename col variants -> variant

    more consistent and concise

  fix variable naming

  move distinct() to proper place

    now actually gets rid of duplicated cols
    added a note on when mutations may be duplicate

  correct comment

  format deconvolution.R

  add * to possible nuc and aa mut chars

    used to denote deletion

  fix case of Others group mixed into dedupe groups

    in that case the lookup of the proportional weight in the
    msig_deduped_df creation step would fail and create an error

  unify execute deconvolution conditions

  make separator for aa mutations a variable

    also change aa mut and nucleotide mut separator to be safer during
    file reading

  remove unused detectable_deletions function

    I looked through the entire results section of the project dir
    and could find no files with "deletions" in them...

  remove calculation of lowecase others col

    This was supposed to already be gone before I worked on this.
    It was commented out, but when I copied stuff around I didn't
    realize it and activated it again.

  format deconvolution_funs.R

  add proper handling of indels to get_protein_mut

    This replaces the detectable_deletions function and works both for
    insertions and deletions. Plus it actually runs.

  fix variable in deconvolution_funs.R

    Missed it during renaming.

  update regex for nuc muts to include indels

    It is conjecture whether this works, but if amino acids have - for
    deletions, nucleotides probably do too.

  remove browser() debugging call i forgot

  fix concat of replacement rows in get_protein_mut

    Missed this earlier, didnt even test it then.

  fix deletion of indel rows in get_protein_mut

jonasfreimuth committed:
prettify debug printing of report parameters

jonasfreimuth committed:
change file io to use fread

  add loading of data.table package where necessary and
  carry out most other related changes

convert read.csv -> fread in index.Rmd

  missed them the first time round

change missed read.csv -> fread

remove debug args loading

  that should not have been committed in the first place

remove additional args to fread

  this makes the corresponding summary function safer as it will throw
  an error when one would supply files with mismatched dfs

add data.table to packages explicitly

  it should already be a dependency of at least one other package

fix package name

add data.table to configure.ac

convert io to fread in qc_rep

  I cant believe I keep finding places I missed

jonasfreimuth committed:
improve creating_*summary_table.R scripts

give input files to create*summary from snakefile

  seems like this was already intended but not followed through with

add newline to log cat statements

format var & mut summary scripts

remove leftover debug comment

remove unused output file arg from create_summary

fix headers

  due to the switch to fread the headers are not changed anymore

jonasfreimuth committed:
adapt index.Rmd and muation_regression.R

set location for mutation summary to mutation dir

  I think I didnt understand it well enough when I put it there

fix and format pool_by_weigthed_mean function

fix run_mutation_regression rule logging

add pretty parameter printing mutation_regression

move function loading to start of Rmd

fix "Others" in meta_cols

move not enough variants warning to proper place

remove outdated TODO

add function for ordering factors for plotting

  and apply it directly

add parsing out of potential single quotes

  this became necessary due to the switch to fread

jonasfreimuth committed:
remove defaults used for debugging in deconvolution

add gene name to nuc_aa_mut string

  done because else this info will be missing in the report

remove unnecessary expandy dots

jonasfreimuth committed:
adapt & refactor mutation_regression.R

simplify parsing_mutation_plot_data

move parsing_data_plot_data code to script

  this should improve understandability, as the function was only used
  once and is not too complicated any more

format & correct names in refined_lm_model

simplify refined_lm_model fun

lms: remove provisional coercion of input to df

add filter for at least 20% of dates covered

jonasfreimuth committed:
adapt index.Rmd

add missing table extraction function

add var_sep to index.Rmd

fix mutation_sheet reading in index.Rmd

fix gene name removal from sigmuts_df

fix sigflag creation in index.Rmd

fix html_note creation

add dplyr package to fun_index.Rmd

  resolves
  https://github.com/BIMSBbioinfo/pigx_sars-cov-2/pull/121#discussion_r884171332

remove resolved TODO on Others being first

fix index.Rmd plots & finish sigflag procedure

add coercion of dates to character in location plot

  necessary as otherwise the plot uses each data point twice for some
  reason

apply sigmut procedure to mutation x location plot

  Add TODO for unifying the sigflag vals

add date conversion after sorting to mut plot

  same reason as with variant plot

add formating of message describing mut string

change variable names to snake_case

enforce line width

adapt nuc_aa_mut string explanation properly

Fix typo in index.Rmd

Fix SigFlag computation in index.Rmd

  Previously, if  a mutation was not a sigmut it generated an
  replacement has length zero error

jonasfreimuth committed:
pass deconvolution output paths from snakefile

  remove hardcoded locations in script
  see issue #114

jonasfreimuth committed:
format deconvolution.R

alexg9010 commented on scripts/creating_variants_summary_table.R:
```diff
@@ -1,41 +1,40 @@
+output_file <- args[1]
+files       <- args[2:length(args)]
+
+output <- create_summary(files)
```
The whole processing within the scripts `creating_variants_summary_table.R` and `creating_mutation_summary_table.R` is identical, except for suffices variants vs mutations. 
Since the input files and output filename are given by arguments, it might be a good idea to join both scripts into a `creating_variants_summary_table.R` script?

jonasfreimuth replied:
Done

alexg9010 commented on scripts/deconvolution_funs.R:
```diff
@@ -0,0 +1,255 @@
+library("dplyr")
+library("data.table")
+
+parse_snv_csv <- function(snvfile, ...) { # allele frequency from v-pipe vcf
```
why do you need these `...` as arguments? They are not passed to any other downstream function ...

jonasfreimuth replied:
Removed it

alexg9010 commented on scripts/deconvolution_funs.R:
```diff
@@ -0,0 +1,255 @@
+  #' caller
+  #' it was build by Altuna
+
+  rlm_model <- suppressWarnings(MASS::rlm(sig, bulk, maxit = 100, method = "M"))
```
Is the `MASS` R-library listed as a dependency?

jonasfreimuth replied:
Well, according to Dirk Eddelbuettel in 2024/16 (https://stackoverflow.com/a/9705725) it is at least in the recommended packages. But to be safe I will explicitly load it and include it in the manifest and other build files...

jonasfreimuth replied:
See the lines [here ](https://github.com/BIMSBbioinfo/pigx_sars-cov-2/pull/121/files#diff-49473dca262eeab3b4a43002adb08b4db31020d190caaad1594b47f1d5daa810R69), [here](https://github.com/BIMSBbioinfo/pigx_sars-cov-2/pull/121/files#diff-f2d3269a9b49cddd2b846cf38ebbae68a5d7bbadb390668e38c5c3a773b599d1R1), and [here](https://github.com/BIMSBbioinfo/pigx_sars-cov-2/pull/121/files#diff-4a3434b9932613220180f203e1970bf52b99ab6a4d66fc0bcdde9e8a8060bc68R66)

alexg9010 commented on scripts/pred_mutation_increase.R:
```diff
@@ -1,84 +1,45 @@
+  # to get the change of each mutations abundance over all dates
+  all_lm_res <- lapply(
+    mutations,
+    function(mutation) {
+      mutation_col <- mutations_df[[mutation]]
+
+      # only do regression if >= 20% of dates have detected mutations
+      # (if not, this will return NULL)
+      if (sum(mutation_col > 0) >= 0.2 * length(dates)) {
+        lm      <- lm(mutation_col ~ dates)
+        lm_sum  <- summary(lm)
+
+        data.frame(
+          mutation     = mutation,
+          coefficients = lm_sum$coefficients["dates", "Estimate"],
+          pvalues      = lm_sum$coefficients["dates", "Pr(>|t|)"]
+        )
```
Nice job summarizing this code.  👍 
I hope you verified that the output is identical to the previous code.

jonasfreimuth replied:
I have verified that the function itself outputs exactly the same object when running on the `sepjan_new` "benchmark" dataset as the previous version.

alexg9010 commented on snakefile.py:
```diff
@@ -37,14 +37,14 @@
 READ_LENGTH      = config['trimming']['read-length']
 CUT_OFF          = config['trimming']['cut-off']
 
+MUTATION_DEPTH_THRESHOLD    = config["reporting"]["mutation-depth-threshold"]
```
I saw where this is being used, but what makes it different from `MUTATION_COVERAGE_THRESHOLD`?

jonasfreimuth replied:
MUTATION_COVERAGE_THRESHOLD gives the percentage of mutations in the mutation_sheet that need to be covered by each sample to be used in the mutation regression, whereas MUTATION_DEPTH_THRESHOLD gives the minumum read coverage at the position of each mutation that needs to be met in order for the mutation to be considered as a signature mutation... I suppose I should mention that somewhere...

alexg9010 commented on scripts/report_scripts/variantreport_p_sample.Rmd:
```diff
@@ -27,468 +25,182 @@ params:
+matching the provided list of lineage - characterizing mutations (signature 
+mutations). For detailed method description please consult the
+[documentation](http://bioinformatics.mdc-berlin.de/pigx_docs/pigx-sars-cov-2.html#output-description)
+*Note: Deconvolution results are biased by the number of reads and by the number
+of signature mutations that were found for this sample.  Check with the QC 
+report to be able to assess how trustworthy the results are.*
```
Is this sentence still rendered as italic?

jonasfreimuth replied:
It is. Only the markup in the code looks different. But apparently as long as there is a `*` next to the last non whitespace character of the block it is fine...

jonasfreimuth committed:
merge create_*_summary_table.R scripts

adapt variable naming in create_summary_table.R

simplify sorting out of empty tables in create sum

adapt snakefile to create_summary_table.R

adapt Makefile.am to create_summary_table.R

fix create_summary_table.R file  name

format create_mutations_summary rule

format create_variants_summary_rule

jonasfreimuth committed:
make MASS package use explicit

add explicit MASS package loading to deconv funs

add MASS package to build related files

jonasfreimuth committed:
remove "CODE" button on reports

jonasfreimuth committed:
adapt qc_report_per_sample

rename coverage.df -> coverage_df in qc_report

rename variables to snake_case in qc_report

add removal of "'" from not covered muts in qc_rep

rework generation of tables for qc_rep

add fixme for colnames in qc_rep

add setting of qc_rep dl out file from snakefile

  this also fixes a potential older issue of the filename of the output
  file not being unique across samples and thus being overwritten

fix missing data.table library in qc_rep

format render_qc_report_rule

jonasfreimuth committed:
point automake flag to pigx-common m4 dir

  otherwise alocal complains about not finding the m4 dir with a warning
  there didn't seem to be any other consequences

jonasfreimuth committed:
change snakefile print calls to log calls

  Printed output can lead to trouble, like the generation of dag output
  not working due to the snakemake output being polluted by print output
  using the logger ensures that output goes where snakemake can use it

jonasfreimuth committed:
update r packages

  remove tidyverse package from manifest.scm

    tidyverse bundles a lot of packages and so constitutes a lot of
    ballast. All the packages we need are already mentioned explicitly.
    Therefore we remove this package to reduce overhead.

  add prv untracked packages to manifest & configure

  sort r pckgs in configure & manifest by alphabet

  substitute required packages only for tidyverse

    Removes tidyverse from Sample_prep_scripts/sigmuts_on_amplicons.Rmd

jonasfreimuth commented on scripts/deconvolution_funs.R:
```diff
@@ -0,0 +1,179 @@
+    # is still better, because due to a bug the previous solution could almost
+    # never run. Now there is at least a warning...
+    if (any(!is.na(locations$mut_end))) {
+      warning("Found a deletion. Help.")
```
The dangling issue of how to deal with deletions here is fixed on my local branch btw. I will push once the review is through, or I am prompted to do so.

jonasfreimuth replied:
Oops, during some reshuffling of my local repo I didn't realize this PR would be affected and pushed onto the branch. The result should be that commit "simplify deconvolution" now has additional subcommits relating to this point, see below.
```
  remove unused detectable_deletions function

    I looked through the entire results section of the project dir
    and could find no files with "deletions" in them...

  remove calculation of lowecase others col

    This was supposed to already be gone before I worked on this.
    It was commented out, but when I copied stuff around I didn't
    realize it and activated it again.

  format deconvolution_funs.R

  add proper handling of indels to get_protein_mut

    This replaces the detectable_deletions function and works both for
    insertions and deletions. Plus it actually runs.

  fix variable in deconvolution_funs.R

    Missed it during renaming.

  update regex for nuc muts to include indels

    It is conjecture whether this works, but if amino acids have - for
    deletions, nucleotides probably do too.

  remove browser() debugging call i forgot

  fix concat of replacement rows in get_protein_mut

    Missed this earlier, didnt even test it then.

  fix deletion of indel rows in get_protein_mut
```
